drafts/records.md

```diff
@@ -0,0 +1,398 @@
+
+Calling `clone` on a `record` results in the same record object being returned. As it is a "value" object, it
+represents a value and is the same thing as saying `clone 3`—you expect to get back a `3`.
+
+`with` may be called with no arguments, and it is the same behavior as `clone`.
+This is an important consideration because a developer may call `$new = $record->with(...$array)` and we don’t want to
+crash.
+If a developer wants to crash, they can do by `assert($new !== $record)`.
+
+### Serialization and deserialization
+
+Records are fully serializable and deserializable.
+
+```php
+record Single(string $value);
+record Multiple(string $value1, string $value2);
+
+echo $single = serialize(Single('value')); // Outputs: "O:6:"Single":1:{s:5:"value";s:5:"value";}"
+echo $multiple = serialize(Multiple('value1', 'value2')); // Outputs: "O:8:"Multiple":1:{s:6:"values";a:2:{i:0;s:6:"value1";i:1;s:6:"value2";}}"
+
+echo unserialize($single) === Single('value'); // Outputs: true
+echo unserialize($multiple) === Multiple('value1', 'value2'); // Outputs: true
+```
+
+### Equality
+
+A `record` is always strongly equal (`===`) to another record with the same value in the properties,
+much like an `array` is strongly equal to another array containing the same elements.
+For all intents, `$recordA === $recordB` is the same as `$recordA == $recordB`.
+
+Comparison operations will behave exactly like they do for classes, for example:
+
+```php
+record Time(float $milliseconds = 0) {
+    public float $totalSeconds {
+        get => $this->milliseconds / 1000,
+    }
+    
+    public float $totalMinutes {
+        get => $this->totalSeconds / 60,
+    }
+    /* ... */
+}
+
+$time1 = Time(1000);
+$time2 = Time(5000);
+
+echo $time1 < $time2; // Outputs: true
+```
+
+### Reflection
+
+Records in PHP will be fully supported by the reflection API,
+providing access to their properties and methods just like regular classes.
+However, immutability and special instantiation rules will be enforced.
+
+#### ReflectionClass support
+
+It can be used to inspect records, their properties, and methods. Any attempt to modify record properties
+via reflection will throw an exception, maintaining immutability. Attempting to create a new instance via
+`ReflectionClass` will cause a `ReflectionException` to be thrown.
+
+``` php
+$point = Point(3, 4);
+$reflection = new \ReflectionClass($point);
+
+foreach ($reflection->getProperties() as $property) {
+    echo $property->getName() . ': ' . $property->getValue($point) . PHP_EOL;
+}
+```
+
+#### Immutability enforcement
+
+Attempts to modify record properties via reflection will throw an exception.
+
+``` php
+try {
+    $property = $reflection->getProperty('x');
+    $property->setValue($point, 10); // This will throw an exception
+} catch (\ReflectionException $e) {
+    echo 'Exception: ' . $e->getMessage() . PHP_EOL; // "Cannot modify a record property"
+}
+```
+
+#### ReflectionFunction for implicit constructor
```

Comment:
Ah, _"implicit constructor."_ Nice!

But what happens when there is an _Explicit_ constructor?